fix(provider/aws): Remove copySourceSpotPrice property

tanelso2 opened the pull request:
This property is no longer needed and was causing Spinnaker to deploy
serverGroups with spotPrices even though the spot price had been removed
(because a previous deployment had the spot price set)

Related to https://github.com/spinnaker/deck/pull/4043 and https://github.com/spinnaker/orca/pull/1572

duftler commented:
This change introduced a regression for users with existing pipelines containing clone stages. All of the other attributes of the source server group being cloned are copied over. Why is this particular attribute not being copied over anymore?

There are both ad-hoc and pipeline Clone operations.

In an ad-hoc Clone (or Deploy), the user has the opportunity to explicitly specify a spot price. But in a pipeline Clone stage, everything (except for the image being updated) is inherited from the source server group. With this change, this is no longer the behavior.

tanelso2 commented:
Ah shoot sorry I didn't mean to break things. 

I removed this because when I had a server group with a spot price set in the UI, and I removed the spot price, it would still get deployed with the old spot price due to this property.

duftler commented:
I think what we've typically done in cases like this is the following:
- Clone request contains an explicit value -> use that value
- Clone request has no explicit attribute value -> use inherited value
- Clone request specifies null -> don't set value on new server group (that is, use whatever the default is)

I think if we undo this change, and ensure the ui doesn't send `null` when the spot price is removed, we'll have the correct behavior all-around.

@anotherchrisberry WDYT?

tanelso2 commented:
We'd need to change deck as well, because the spotPrice field is sending back `null` when it is unset because AWS was failing when sent an empty string

duftler commented:
Yeah, I agree we need a change in both components.

anotherchrisberry commented:
All seems reasonable to me.

duftler commented:
Frontend change is here: https://github.com/spinnaker/deck/pull/4114
Backend change is here: https://github.com/spinnaker/clouddriver/pull/1908
